tests/run-custom-args/captures/minicheck.scala

```diff
@@ -0,0 +1,230 @@
+
+class Run:
+  var errorCount = 0
+
+object report:
+  def error(msg: -> String)(using Context) =
+    ctx.run.errorCount += 1
+    println(s"ERROR: $msg")
+
+abstract class Ctx:
+  def outer: Context
+  def owner: Symbol
+  def scope: Scope
+  def run: Run
+  def detached: DetachedContext
+
+type Context = {*} Ctx
+
+abstract class DetachedContext extends Ctx:
+  def outer: DetachedContext
+
+class FreshCtx(val level: Int) extends DetachedContext:
+  var outer: FreshCtx = uninitialized
+  var owner: Symbol = uninitialized
+  var scope: Scope = uninitialized
+  var run: Run = uninitialized
+  def initFrom(other: Context): this.type =
+    outer = other.asInstanceOf[FreshCtx]
+    owner = other.owner
+    scope = other.scope
+    run = other.run
+    this
+  def detached: DetachedContext =
+    var c = this
+    while c.level >= 0 && (ctxStack(c.level) eq c) do
+      ctxStack(c.level) = FreshCtx(c.level)
+      c = c.outer
+    this
+
+object NoContext extends FreshCtx(-1):
+  owner = NoSymbol
+  scope = EmptyScope
+
+type FreshContext = {*} FreshCtx
+
+def ctx(using c: Context): {c} Ctx = c
+
+// !cc! it does not work if ctxStack is an Array[FreshContext] instead.
+var ctxStack = Array.tabulate(16)(new FreshCtx(_))
+var curLevel = 0
+
+def inFreshContext[T](op: FreshContext ?-> T)(using Context): T =
+  if curLevel == ctxStack.length then
+    val prev = ctxStack
+    ctxStack = new Array[FreshCtx](curLevel * 2)
+    Array.copy(prev, 0, ctxStack, 0, prev.length)
+    for level <- curLevel until ctxStack.length do
+      ctxStack(level) = FreshCtx(level)
+  val result = ctxStack(curLevel).initFrom(ctx)
+  curLevel += 1
+  try op(using result)
+  finally curLevel -= 1
+
+def withOwner[T](owner: Symbol)(op: Context ?-> T)(using Context): T =
+  val prev = ctx
+  inFreshContext: c ?=>
+    c.owner = owner
+    op
+
+def withScope[T](scope: Scope)(op: Context ?-> T)(using Context): T =
+  val prev = ctx
+  inFreshContext: c ?=>
+    c.scope = scope
+    op
+
+def typed(tree: Tree, expected: Type = NoType)(using Context): Type =
+  try
+    val tp = typedUnadapted(tree, expected)
+    if expected.exists && tp != expected then
+      report.error(
+        s"""Type error
+          |  found   : $tp
+          |  expected: $expected
+          |  for     : $tree""".stripMargin)
+    tp
+  catch case ex: TypeError =>
+    report.error(ex.msg)
+    NoType
+
+import Tree.*
+def typedUnadapted(tree: Tree, expected: Type = NoType)(using Context): Type = tree match
+  case Let(bindings, res) =>
+    withScope(Scope()):
+      for Binding(name, rhs) <- bindings do
+        val sym = Symbol(name, ctx.owner)
+        val dctx = ctx.detached
+        sym.info = new LazyType(using dctx):
+          override def doComplete()(using Context) =
+            sym.info = withOwner(sym):
+              typed(rhs)
+        ctx.scope.enter(sym)
+      try typed(res, expected)
+      finally for sym <- ctx.scope.elements do sym.info
```

Comment:
Just out of curiosity, what's the difference if you force the bindings first and the result after?  Is there a reason to do it in this order?

Reply:
The way it is done allows forward references in bindings.

Reply:
In fact, I think I misunderstood the question at first. In the try it could be done either way. And in fact in the compiler it's done in the opposite order.